packages/uploadthing/src/internal/ut-reporter.ts

```diff
@@ -0,0 +1,65 @@
+import { UploadThingError } from "@uploadthing/shared";
+
+import { maybeParseResponseXML } from "./s3-error-parser";
+import type { ActionType, UTEvents } from "./types";
+
+export const createAPIRequestUrl = (config: {
+  url?: string;
+  slug: string;
+  actionType: ActionType;
+}) => {
+  const url = new URL(
+    config.url ?? `${window.location.origin}/api/uploadthing`,
+  );
+
+  const queryParams = new URLSearchParams(url.search);
+  queryParams.set("actionType", config.actionType);
+  queryParams.set("slug", config.slug);
+
+  url.search = queryParams.toString();
+  return url.toString();
+};
+
+/**
+ * Creates a "client" for reporting events to the UploadThing server via the user's API endpoint.
+ * Events are handled in "./handler.ts starting at L200"
+ */
+export const createUTReporter = (cfg: { url?: string; endpoint: string }) => {
+  return async <TEvent extends keyof UTEvents>(
+    type: TEvent,
+    payload: UTEvents[TEvent],
+  ) => {
+    const url = createAPIRequestUrl({
+      url: cfg.url,
+      slug: cfg.endpoint,
+      actionType: type,
+    });
+    const response = await fetch(url, {
+      method: "POST",
+      body: JSON.stringify(payload),
+      headers: { "Content-Type": "application/json" },
```

Comment:
should we include the version and key here?

Reply:
This just goes to the users backend so we dont have the key here